OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -0,0 +1,324 @@
+                                   this,
+                                   context,
+                                   stream,
+                                   &all_events_sent_received,
+                                   &exit_send_commands_thread};
+
+  // This thread is responsible for reading from stream, and specifically for
+  // receiving CaptureEvents and AllEventsSent messages.
+  std::thread receive_events_thread{&ProducerSideServiceImpl::ReceiveEventsThread, this, context,
+                                    stream, &all_events_sent_received};
+  receive_events_thread.join();
+
+  // When receive_events_thread exits because stream->Read(&request) fails,
+  // it means that the producer has disconnected: ask send_commands_thread to exit, too.
+  exit_send_commands_thread = true;
+  send_commands_thread.join();
+
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    server_contexts_.erase(context);
+  }
+
+  LOG("Finished handling ReceiveCommandsAndSendEvents for a CaptureEventProducer");
+  return grpc::Status::OK;
+}
+
+void ProducerSideServiceImpl::SendCommandsThread(
+    grpc::ServerContext* context,
+    grpc::ServerReaderWriter<orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse,
+                             orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream,
+    bool* all_events_sent_received, std::atomic<bool>* exit_send_commands_thread) {
+  // As a result of initializing prev_capture_status to kCaptureFinished,
+  // an initial StartCaptureCommand is sent
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStarted,
+  // and an initial StopCaptureCommand is sent (with little effect)
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStopping.
+  CaptureStatus prev_capture_status = CaptureStatus::kCaptureFinished;
+
+  // This loop keeps track of changes to service_state_.capture_status using
+  // conditional critical sections on service_state_mutex_ and updating prev_capture_status,
+  // and sends StartCaptureCommands and StopCaptureCommands accordingly.
+  // It exits when *exit_send_commands_thread is true, when service_state_.exit_requested is true,
+  // or when Write fails (because the producer disconnected or because the context was cancelled).
+  while (true) {
+    // This is set when ReceiveEventsThread has exited. At that point this thread should also exit.
+    if (*exit_send_commands_thread) {
+      return;
+    }
+
+    service_state_mutex_.Lock();
+    if (service_state_.exit_requested) {
+      service_state_mutex_.Unlock();
+      return;
+    }
+
+    if (service_state_.capture_status != prev_capture_status) {
+      switch (service_state_.capture_status) {
+        case CaptureStatus::kCaptureStarted: {
+          prev_capture_status = service_state_.capture_status;
+          ++service_state_.producers_remaining;
+          *all_events_sent_received = false;
+          service_state_mutex_.Unlock();
```

Comment:
Yes this is really hard to track. One thing you could do is to set prev_capture_status in L153, You can then use it in switch without locking. As for other fields... What would happen if you reistablished the lock here? Is there a possibility of race condition on them in other words do they have to be scoped with status read?

Reply:
Hmmm, I don't think the mutex should be acquired twice, as `service_state_.capture_status` and `service_state_.producers_remaining` and also `all_events_sent_received` go together. The value assigned to the last two depends quite strongly on the current value of `service_state_.capture_status`. That shouldn't be relaxed only to make it easier to track `Unlock`s.

Reply:
ok, then split the case into 2. One scoped on the mutex - and the second one doing communication based on prev_capture_status, it is going to be more readable and will have a clear mutex scope.

Reply:
You mean I should have two separate `switch`es?

Reply:
yes, switches, it makes code more readable, and mutex scope easier to follow.

Reply:
Makes sense, done.